### DISCUSSION TITLE
[General] proveL2MessageInclusion reverted with reason "rz"

### DISCUSSION THREAD

**Discussion by Sam-Scolari** [zkSync-Community-Hub/zksync-developers#211] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to send and prove an L2 -> L1 message

L2 Contract: 0x04d51E91c689B4cf573977d79631076609fD0d14
TX Hash: [0xbdec3d890fef30a0699d59814997e98dfa156479ee6ab67420de8bc8eae80178](https://goerli.explorer.zksync.io/tx/0xbdec3d890fef30a0699d59814997e98dfa156479ee6ab67420de8bc8eae80178#eventlog)

### What is the specific issue or error you're encountering?

Using the code above calling proveL2MessageInclusion reverts with reason "rz". 

### Can you share the error messages or logs you're receiving, if any?

```js
{ reason: 'rz',
  code: 'CALL_EXCEPTION',
  method: 'proveL2MessageInclusion(uint256,uint256,(uint16,address,bytes),bytes32[])',
  data: '0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002727a000000000000000000000000000000000000000000000000000000000000',
  errorArgs: [ 'rz' ],
  errorName: 'Error',
  errorSignature: 'Error(string)',
  address: '0x1908e2bf4a88f91e4ef0dc72f02b8ea36bea2319',
  args: [
    221191,
    0,
    {
      txNumberInBlock: 6,
      sender: '0x04d51E91c689B4cf573977d79631076609fD0d14',
      data: '0x0000000000000000000000000000000000000000000000000000000000000045000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000002'
    },
    [
      '0xc08d04b5eb488bd75830b43b89398aa26f78090df33b2fec82900e03bbe3b5d4',
      '0xc3d03eebfd83049991ea3d3e358b6712e7aa2e2e63dc2d4b438987cec28ac8d0',
      '0xe3697c7f33c31a9b0f0aeb8542287d0d21e8c4cf82163d0c44c7a98aa11aa111',
      '0x199cc5812543ddceeddd0fc82807646a4899444240db2c0d2f20c3cceb5f51fa',
      '0xe4733f281f18ba3ea8775dd62d2fcd84011c8c938f16ea5790fd29a03bf8db89',
      '0x1798a1fd9c8fbb818c98cff190daa7cc10b6e5ac9716b4a2649f7c2ebcef2272',
      '0x66d7c5983afe44cf15ea8cf565b34c6c31ff0cb4dd744524f7842b942d08770d',
      '0xb04e5ee349086985f74b73971ce9dfe76bbed95c84906c5dffd96504e1e5396c',
      '0xac506ecb5465659b3a927143f6d724f91d8d9c4bdb2463aee111d9aa869874db',
      '0x124b05ec272cecd7538fdafe53b6628d31188ffb6f345139aac3c3c1fd2e470f',
      '0xc3be9cbd19304d84cca3d045e06b8db3acd68c304fc9cd4cbffe6d18036cb13f'
    ]
  ],
  transaction: {
    data: '0xe4948f430000000000000000000000000000000000000000000000000000000000036007000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000180000000000000000000000000000000000000000000000000000000000000000600000000000000000000000004d51e91c689b4cf573977d79631076609fd0d14000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000045000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000bc08d04b5eb488bd75830b43b89398aa26f78090df33b2fec82900e03bbe3b5d4c3d03eebfd83049991ea3d3e358b6712e7aa2e2e63dc2d4b438987cec28ac8d0e3697c7f33c31a9b0f0aeb8542287d0d21e8c4cf82163d0c44c7a98aa11aa111199cc5812543ddceeddd0fc82807646a4899444240db2c0d2f20c3cceb5f51fae4733f281f18ba3ea8775dd62d2fcd84011c8c938f16ea5790fd29a03bf8db891798a1fd9c8fbb818c98cff190daa7cc10b6e5ac9716b4a2649f7c2ebcef227266d7c5983afe44cf15ea8cf565b34c6c31ff0cb4dd744524f7842b942d08770db04e5ee349086985f74b73971ce9dfe76bbed95c84906c5dffd96504e1e5396cac506ecb5465659b3a927143f6d724f91d8d9c4bdb2463aee111d9aa869874db124b05ec272cecd7538fdafe53b6628d31188ffb6f345139aac3c3c1fd2e470fc3be9cbd19304d84cca3d045e06b8db3acd68c304fc9cd4cbffe6d18036cb13f',
    to: '0x1908e2BF4a88F91E4eF0DC72f02b8Ea36BEa2319'
  }
}
```

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

zksync-web3

### Can you provide the relevant portions of your contract code where the issue is occurring?

Contract:

```solidity
 function ping(address _messenger) external returns(bytes32) {
        IL1Messenger messenger = IL1Messenger(_messenger);

        Message memory message = Message({
            proposal: 69,
            forVotes: 5,
            againstVotes: 3,
            abstainVotes: 2
        });

        return messenger.sendToL1(abi.encode(message));
    }
```

JS Script:

```js
//@ts-check

const { Provider, utils } = require("zksync-web3");
const { getProvider, getWallet } = require("./utils");
const { ethers } = require("ethers");
require("dotenv").config();

const l2ContractAddress = "0x04d51E91c689B4cf573977d79631076609fD0d14";
const l2TransactionHash =
  "0xbdec3d890fef30a0699d59814997e98dfa156479ee6ab67420de8bc8eae80178";

async function main() {
  const l1Provider = new ethers.providers.StaticJsonRpcProvider(
    process.env.GOERLI_RPC_URL
  );
  const zkSyncProvider = await getProvider();

  const message = ethers.utils.AbiCoder.prototype.encode(
    ["uint256", "uint256", "uint256", "uint256"],
    [69, 5, 3, 2]
  );
  console.log("message", message);

  // const message =
  //   "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000045000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000002";
  const messageHash = ethers.utils.keccak256(message);
  // Actual hash: "0x33ff89b6c9cbbfff9b4c4f9fb6b26365ce4b2ffc71f983524625c52f0615ac17"
  console.log("hash", messageHash);

  console.log("Waiting for L1 block inclusion (this may take up to 1 hour)...");

  const { l1BatchNumber, l1BatchTxIndex, blockNumber } =
    await zkSyncProvider.getTransactionReceipt(l2TransactionHash);
  if (l1BatchNumber) {
    const zkAddress = await zkSyncProvider.getMainContractAddress();
    const sender = l2ContractAddress;
    const proofInfo = await zkSyncProvider.getMessageProof(
      blockNumber,
      sender,
      messageHash
    );
    if (!proofInfo) {
      throw new Error("No proof found");
    }
    const index = proofInfo.id;
    const proof = proofInfo.proof;

    const mailboxL1Contract = new ethers.Contract(
      zkAddress,
      utils.ZKSYNC_MAIN_ABI,
      l1Provider
    );

    // all the information of the message sent from L2
    const messageInfo = {
      txNumberInBlock: l1BatchTxIndex,
      sender,
      data: message,
    };

    try {
      const result = await mailboxL1Contract.proveL2MessageInclusion(
        l1BatchNumber,
        index,
        messageInfo,
        proof
      );
      console.log("L2 block:", blockNumber);
      console.log("L1 Index for Tx in block:", l1BatchTxIndex);
      console.log("L1 Batch for block: ", l1BatchNumber);
      console.log("Inclusion proof:", proof);
      console.log("proveL2MessageInclusion:", result);
    } catch (err) {
      console.error(err);
    }
  }
}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
```

### Have you tried to isolate the problem, and if so, what were the results?

The [sendToL1](https://github.com/matter-labs/era-system-contracts/blob/main/contracts/L1Messenger.sol#L112) code in the era contracts repo takes the message bytes 

```solidity
function sendToL1(bytes calldata _message) external override returns (bytes32 hash) {
```

computes a keccak256 hash
```solidity
hash = EfficientCall.keccak(_message);
```

and then emits that data as a log.
```solidity
emit L1MessageSent(msg.sender, hash, _message);
```

Looking at [the log](https://goerli.explorer.zksync.io/tx/0xbdec3d890fef30a0699d59814997e98dfa156479ee6ab67420de8bc8eae80178#eventlog) we can see that the data returned is as follows

**Data:** `0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000045000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000002`
**Hash:** `0x33ff89b6c9cbbfff9b4c4f9fb6b26365ce4b2ffc71f983524625c52f0615ac17`

This is not correct as the keccak256 hash for that data is: `0xadf9f8faf9adf7c7ecc94317838d0c411c790e828e33da99822ac1f61df81b05`

However, removing the bytes related to internal struct stuff does in fact produce that hash

**Data:**
`0x0000000000000000000000000000000000000000000000000000000000000045000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000002`

**Hash:** `0x33ff89b6c9cbbfff9b4c4f9fb6b26365ce4b2ffc71f983524625c52f0615ac17`

I am not sure if this inconsistency is the source of the issue or not.

### What steps have you already taken to try to resolve the issue?

I've tried using both data and keccak hash pairs without sucesss

**Comment by montekki** [comment, marked as answer]
Hi @Sam-Scolari could you please re-request the proofs and resubmit? There has been a bug in the API that has returned wrong proof lengths, so in your case the length of 11:

```
[
      '0xc08d04b5eb488bd75830b43b89398aa26f78090df33b2fec82900e03bbe3b5d4',
      '0xc3d03eebfd83049991ea3d3e358b6712e7aa2e2e63dc2d4b438987cec28ac8d0',
      ...
```

and the contract reverts with this reason. If you request this proof now it should be an array of length 9 and then it should work. LMK if that helps.

**Comment by keegansa** [comment]
It seems like the issue might be related to the format or encoding of the message being sent from L2 to L1. The discrepancy between the expected hash and the actual hash suggests that the message data might not be formatted correctly before hashing.

To resolve this issue, ensure that the message data is correctly encoded according to the expected format on both the L2 and L1 sides. Double-check the encoding and decoding logic to make sure there are no discrepancies or missing steps.

Additionally, verify that the proof generation and inclusion process is implemented correctly and that all necessary parameters are provided accurately to the proveL2MessageInclusion function.

If the problem persists, consider reaching out to the zkSync support team or community for further assistance, as they may have insights or documentation that can help troubleshoot and resolve the issue.

**Comment by marqbassmier** [comment]
Based on the provided information, it seems that the inconsistency in the keccak hash computation might be causing the issue. Ensure that the message being sent to L1 and its corresponding hash are computed correctly. Double-check the implementation of the sendToL1 function in the L2 contract to ensure it aligns with your expectations.

**Comment by meutrandeesni** [comment]
It appears that the data being passed to proveL2MessageInclusion might not be correctly formatted or computed. Double-check the data being passed, including the message format and its hash, to ensure consistency with the contract's expectations. Additionally, verify that the proof generation process is accurate and includes all necessary information for successful validation.

**Comment by porbotufibr** [comment]
The revert reason "rz" indicates that there was an error during the execution of the proveL2MessageInclusion function. Check the function implementation and ensure that all conditions and requirements are met for successful execution.
